packages/vitest-pool-workers/src/pool/index.ts

```diff
@@ -0,0 +1,767 @@
+				) {
+					startupError = value.vitestPoolWorkersError;
+				} else {
+					listener(value);
+				}
+			});
+		},
+	});
+	project.project.ctx.onCancel((reason) => rpc.onCancel(reason));
+	webSocket.accept();
+
+	const [event] = (await events.once(webSocket, "close")) as [CloseEvent];
+	if (webSocket.readyState === WebSocket.READY_STATE_CLOSING) {
+		if (event.code === 1005 /* No Status Received */) {
+			webSocket.close();
+		} else {
+			webSocket.close(event.code, event.reason);
+		}
+	}
+	if (event.code !== 1000) {
+		throw startupError ?? new Error("Failed to run tests");
+	}
+
+	debuglog("DONE", files);
+}
+
+function getRelativeProjectPath(projectPath: string | number) {
+	if (typeof projectPath === "number") return projectPath;
+	else return path.relative("", projectPath);
+}
+
+interface PackageJson {
+	version?: string;
+	peerDependencies?: Record<string, string | undefined>;
+}
+function getPackageJson(dirPath: string): PackageJson | undefined {
+	// eslint-disable-next-line no-constant-condition
+	while (true) {
+		const pkgJsonPath = path.join(dirPath, "package.json");
+		try {
+			const contents = fs.readFileSync(pkgJsonPath, "utf8");
+			return JSON.parse(contents);
+		} catch (e) {
+			if (!isFileNotFoundError(e)) throw e;
+		}
+		const nextDirPath = path.dirname(dirPath);
+		// `path.dirname()` of the root directory is the root directory
+		if (nextDirPath === dirPath) return;
+		dirPath = nextDirPath;
+	}
+}
+
+function assertCompatibleVitestVersion(ctx: Vitest) {
+	// Some package managers don't enforce `peerDependencies` requirements,
+	// so add a runtime sanity check to ensure things don't break in strange ways.
+	const poolPkgJson = getPackageJson(__dirname);
+	const vitestPkgJson = getPackageJson(
+		path.dirname(ctx.projectFiles.workerPath)
+	);
+	assert(
+		poolPkgJson !== undefined,
+		"Expected to find `package.json` for `@cloudflare/vitest-pool-workers`"
+	);
+	assert(
+		vitestPkgJson !== undefined,
+		"Expected to find `package.json` for `vitest`"
+	);
+
+	const expectedVitestVersion = poolPkgJson.peerDependencies?.vitest;
+	const actualVitestVersion = vitestPkgJson.version;
+	assert(
+		expectedVitestVersion !== undefined,
+		"Expected to find `@cloudflare/vitest-pool-workers`'s `vitest` version constraint"
+	);
+	assert(
+		actualVitestVersion !== undefined,
+		"Expected to find `vitest`'s version"
+	);
+
+	if (expectedVitestVersion !== actualVitestVersion) {
+		const message = [
+			`You're running \`vitest@${actualVitestVersion}\`, but this version of \`@cloudflare/vitest-pool-workers\` only supports \`vitest@${expectedVitestVersion}\`.`,
+			"`@cloudflare/vitest-pool-workers` currently depends on internal Vitest APIs that are not protected by semantic-versioning guarantees.",
+			`Please install \`vitest@${expectedVitestVersion}\` to continue using \`@cloudflare/vitest-pool-workers\`.`,
+		].join("\n");
+		throw new Error(message);
+	}
+}
+
+export default function (ctx: Vitest): ProcessPool {
```

Comment:
As far as I can tell from the docs, this function could potentially be called multiple times if the server configuration changes (https://vitest.dev/advanced/pool.html#api). While I don't know how often that's likely to happen, it feels like it would be safer to initialise the spec cache here rather than at the top level, to ensure it's cleared on re-calls?